docs/zkapps/tutorials/05-common-types-and-functions.mdx

```diff
@@ -0,0 +1,360 @@
+
+console.log(`private key: ${privateKey.toBase58()}`);
+console.log(`public key: ${publicKey.toBase58()}`);
+console.log(`Fields in private key: ${privateKey.toFields().length}`);
+console.log(`Fields in public key: ${publicKey.toFields().length}`);
+
+console.log(`signature verified for data1: ${verifiedData1.toString()}`);
+console.log(`signature verified for data2: ${verifiedData2.toString()}`);
+
+console.log(`Fields in signature: ${signature.toFields().length}`);
+```
+
+And the console output:
+
+```sh
+str1: abc..xyz
+Fields in str1: 128
+private key: EKF7vyCyexH6KZpBo4jPG29Ac7tcuYwS1MFwUudhZ6D8v3Ap88SY
+public key: B62qo32qFbpLybPGtFDW8GT5Efu4ZtbUXDoj5vyTrmSp6AXgMh7btAH
+Fields in private key: 255
+Fields in public key: 2
+signature verified for data1: false
+signature verified for data2: true
+Fields in signature: 256
+```
+
+## CircuitValue
+
+:::info
+`CircuitValue` is soon to be replaced with `Struct` in an upcoming version of SnarkyJS. Expect this tutorial to update to use `Struct` in the future accordingly.
+:::
+
+A special type is [CircuitValue](../snarkyjs-reference/classes/CircuitValue), which lets you create your own types for your application.
+
+All arguments passed into smart contracts need to be arguments SnarkyJS can understand. Creating a CircuitValue allows you pass in structured pieces of data, which via their class definition can come with methods to manipulate them. 
+
+See the following for an example of how to use this, implementing a `Points` structure, and an array of points of length 8 structure:
+
+```ts
+class Point extends CircuitValue {
+  @prop x: Field;
+  @prop y: Field;
+
+  static fromField(x: Field, y: Field) {
+    return new Point(x, y);
+  }
+
+  static addPoints(a: Point, b: Point) {
+    return new Point(a.x.add(b.x), a.y.add(b.y));
+  }
+}
+
+const point1 = Point.fromField(Field(10), Field(4));
+const point2 = Point.fromField(Field(1), Field(2));
+
+const pointSum = Point.addPoints(point1, point2);
+
+console.log(`pointSum Fields: ${pointSum.toFields().map((p) => p.toString())}`);
+
+class Points8 extends CircuitValue {
+  @arrayProp(Point, 8) points: Point[];
+
+  constructor(points: Point[]) {
+    super();
+    this.points = points;
+  }
+}
+
+const pointsArray = new Array(8).fill(null).map((_, i) => Point.fromField(Field(i), Field(i*10)));
+const points8 = new Points8(pointsArray);
+
+console.log(`points8 Fields: ${JSON.stringify(points8)}`);
+```
+
+Which prints the following to the console:
+
+```sh
+pointSum Fields: 11,6
+points8 Fields: {"points":[{"x":"0","y":"0"},{"x":"1","y":"10"},{"x":"2","y":"20"},{"x":"3","y":"30"},{"x":"4","y":"40"},{"x":"5","y":"50"},{"x":"6","y":"60"},{"x":"7","y":"70"}]}
+```
+
+A great example of a complex use of CircuitValue is the MerkleTree implementation  - see the source for that [here](https://github.com/o1-labs/snarkyjs/blob/main/src/lib/merkle_tree.ts) for more.
+
+## Control flow
+
+There are two functions which help do control flow within SnarkyJS:
```

Comment:
I think it'd be useful to start this paragraph explaining:

> Because the underlying proof system used by SnarkyJS--Kimchi--does not yet support branching, we cannot use regular `if` statements in our code. But we have two other, circuit-compatible options available to us for flow control:

Reply:
Do we think it ever could support branching? I thought it had to be fixed circuits, cool if this is possible tho

Reply:
My understanding from Izzy a while back is that yes, he thinks there is a way to support true branching someday